### PR TITLE
Issue #5939 Use unwrapped exception as exception type for error handling

### LINKED ISSUE #5939: Use unwrapped exception as exception type for error handling
some context here https://github.com/jetty-project/tck-run/issues/14

### PR DESCRIPTION
**Pull request by olamy**
#5939 
Signed-off-by: olivier lamy <oliver.lamy@gmail.com>

### CODE REVIEW

File: jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ErrorPageErrorHandler.java [hunk 1]
```diff
@@ -75,7 +75,19 @@ public String getErrorPage(HttpServletRequest request)
             if (errorPage != null)
                 matchedThrowable = exClass;
 
-            th = (th instanceof ServletException) ? ((ServletException)th).getRootCause() : null;
+            if (th instanceof ServletException)
```
Comment (janbartel) [inline jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ErrorPageErrorHandler.java:78]: If we change the ERROR_EXCEPTION here, then the LOG.DEBUG generated down at line 141/142 will be wrong.

Also, presumably jetty has had this behaviour for a very long time, so I'm wary about changing it ... @gregw WDYT?

**Review by janbartel — CHANGES_REQUESTED** [review]

File: jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ErrorPageErrorHandler.java (hunk 1 above)
Comment (gregw) [reply jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ErrorPageErrorHandler.java:78]: I think we need to provide a compliance mode to at least be able to switch back to the old behaviour, as we have offered this way for some time.

File: jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ErrorPageErrorHandler.java [hunk 2]
```diff
@@ -75,7 +75,19 @@ public String getErrorPage(HttpServletRequest request)
...
+            {
+                th = ((ServletException)th).getRootCause();
+                if (th != null)
+                {
+                    request.setAttribute(Dispatcher.ERROR_EXCEPTION_TYPE, th.getClass());
```
Comment (gregw) [inline jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ErrorPageErrorHandler.java:83]: I don't think it is correct to do this here as we are walking the hierarchy.
I think the resetting of `Dispatcher.ERROR_EXCEPTION_TYPE` (and probably `Dispatcher.ERROR_EXCEPTION`) should be done below, only if we have an error page AND we are not in the compliance mode.

**Review by gregw — CHANGES_REQUESTED** [review]

**Comment by gregw** [comment]
@olamy Do you want me to add the compliance mode switch for this?

**Comment by gregw** [comment]
Also this needs to be fixed in 10 and merged to 11.  We don't want 11 to diverge from 10

**Comment by gregw** [comment]
replaced by #5958

**Pull request closed without merging**